File: 98. Validate Binary Search Tree/solution1.js
/** Definition for a binary tree node. */
function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
}

/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function (root) {
    let previousValue = -Infinity;
    let flag = true;
    function dfs(node) {
        if (node) {
            dfs(node.left);
            if (node.val <= previousValue) {
                flag = false;
            }
            previousValue = node.val;
            dfs(node.right);
        }
    }
    dfs(root);
    return flag;
};

File: 98. Validate Binary Search Tree/solution1.py
from typing import *

class TreeNode:
    """Definition for a binary tree node."""
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        flag = True
        prev_value = float('-inf')
        def dfs(node):
            nonlocal flag, prev_value
            if node:
                dfs(node.left)
                if node.val <= prev_value:
                    flag = False
                prev_value = node.val
                dfs(node.right)
        dfs(root)

        return flag